File: frontend/beef-and-chicken/beef-and-chicken/src/features/delivery-rush/engine/deliveryRushSimulator.ts
import type {
  DeliveryRushInput,
  DeliveryRushLiveStats,
  DeliveryRushSimulationResult,
} from "../types/deliveryRush.types";
import { generateBlockedLaneMask } from "./deliveryRushObstacles";
import { DeliveryRushRandom } from "./deliveryRushRandom";
import {
  deliveryRushGameRules,
  getDistancePerTick,
  getObstacleInterval,
} from "./deliveryRushRules";

export function simulateDeliveryRush(
  seed: number,
  inputs: readonly DeliveryRushInput[],
): DeliveryRushSimulationResult {
  const liveResult = simulateDeliveryRushUntilTick(
    seed,
    inputs,
    deliveryRushGameRules.totalTicks,
  );

  return {
    score: liveResult.score,
    distance: liveResult.distance,
    avoidedObstacles: liveResult.avoidedObstacles,
    collisionCount: liveResult.collisionCount,
    maxCombo: liveResult.maxCombo,
  };
}

export function simulateDeliveryRushUntilTick(
  seed: number,
  inputs: readonly DeliveryRushInput[],
  completedTicks: number,
): DeliveryRushLiveStats {
  validateDeliveryRushInputs(inputs);
  validateCompletedTicks(completedTicks);

  const random = new DeliveryRushRandom(seed);

  let playerLane: number = deliveryRushGameRules.startingLane;
  let inputIndex = 0;
  let nextObstacleTick: number = deliveryRushGameRules.firstObstacleTick;
  let activeObstacleTick: number | null = null;
  let activeBlockedLaneMask = 0;
  let activeObstacleCollided = false;
  let jumpStartTick: number | null = null;

  let distance = 0;
  let avoidedObstacles = 0;
  let collisionCount = 0;
  let currentCombo = 0;
  let maxCombo = 0;
  let slowdownTicksRemaining = 0;

  for (let tick = 0; tick < completedTicks; tick++) {
    if (inputIndex < inputs.length && inputs[inputIndex].tick === tick) {
      const input = inputs[inputIndex];

      if (input.action === "move") {
        playerLane = clamp(
          playerLane + input.direction,
          0,
          deliveryRushGameRules.laneCount - 1,
        );
      } else {
        jumpStartTick = tick;
      }

      inputIndex += 1;
    }

    const distancePerTick = getDistancePerTick(tick);

    if (slowdownTicksRemaining > 0) {
      distance += Math.max(1, Math.trunc(distancePerTick / 2));
      slowdownTicksRemaining--;
    } else {
      distance += distancePerTick;
    }

    const collisionWindowStartTick =
      nextObstacleTick - deliveryRushGameRules.collisionWindowBeforeTicks;

    if (activeObstacleTick === null && tick === collisionWindowStartTick) {
      activeObstacleTick = nextObstacleTick;

      activeBlockedLaneMask = generateBlockedLaneMask(random, nextObstacleTick);

      activeObstacleCollided = false;
    }

    if (activeObstacleTick === null) {
      continue;
    }

    const collisionWindowEndTick =
      activeObstacleTick + deliveryRushGameRules.collisionWindowAfterTicks;

    const playerLaneMask = 1 << playerLane;

    const isJumping = isDeliveryRushJumpActive(tick, jumpStartTick);

    const isPlayerTouchingObstacle =
      (activeBlockedLaneMask & playerLaneMask) !== 0;

    if (!activeObstacleCollided && isPlayerTouchingObstacle && !isJumping) {
      collisionCount++;
      currentCombo = 0;

      slowdownTicksRemaining = deliveryRushGameRules.collisionSlowdownTicks;

      activeObstacleCollided = true;

      if (collisionCount >= deliveryRushGameRules.maximumCollisions) {
        break;
      }
    }

    if (tick < collisionWindowEndTick) {
      continue;
    }

    if (!activeObstacleCollided) {
      avoidedObstacles++;
      currentCombo++;

      maxCombo = Math.max(maxCombo, currentCombo);
    }

    nextObstacleTick += getObstacleInterval(activeObstacleTick);

    activeObstacleTick = null;
    activeBlockedLaneMask = 0;
    activeObstacleCollided = false;
  }

  const score =
    distance +
    avoidedObstacles * deliveryRushGameRules.avoidedObstaclePoints +
    maxCombo * deliveryRushGameRules.maxComboMultiplier -
    collisionCount * deliveryRushGameRules.collisionPenalty;

  return {
    score: Math.max(0, score),
    distance,
    avoidedObstacles,
    collisionCount,
    currentCombo,
    maxCombo,
  };
}

export function isDeliveryRushJumpActive(
  currentTick: number,
  jumpStartTick: number | null,
): boolean {
  if (jumpStartTick === null) {
    return false;
  }

  return (
    currentTick >= jumpStartTick &&
    currentTick < jumpStartTick + deliveryRushGameRules.jumpDurationTicks
  );
}

export function validateDeliveryRushInputs(
  inputs: readonly DeliveryRushInput[],
): void {
  if (!Array.isArray(inputs)) {
    throw new TypeError("Inputs must be an array.");
  }

  if (inputs.length > deliveryRushGameRules.maximumInputEvents) {
    throw new Error("Too many input events.");
  }

  let previousTick = -1;
  let previousJumpTick: number | null = null;

  for (const input of inputs) {
    if (
      !input ||
      !Number.isInteger(input.tick) ||
      input.tick < 0 ||
      input.tick >= deliveryRushGameRules.totalTicks
    ) {
      throw new Error("Input tick is outside the game duration.");
    }

    if (input.action !== "move" && input.action !== "jump") {
      throw new Error("Input action must be move or jump.");
    }

    if (
      input.action === "move" &&
      input.direction !== -1 &&
      input.direction !== 1
    ) {
      throw new Error("Input direction must be -1 or 1.");
    }

    if (input.tick <= previousTick) {
      throw new Error("Input ticks must be strictly increasing.");
    }

    if (
      previousTick >= 0 &&
      input.tick - previousTick <
        deliveryRushGameRules.minimumTicksBetweenInputs
    ) {
      throw new Error("Input events are too close together.");
    }

    if (input.action === "jump" && previousJumpTick !== null) {
      const nextAllowedJumpTick =
        previousJumpTick +
        deliveryRushGameRules.jumpDurationTicks +
        deliveryRushGameRules.jumpCooldownTicks;

      if (input.tick < nextAllowedJumpTick) {
        throw new Error("Jump is still on cooldown.");
      }
    }

    if (input.action === "jump") {
      previousJumpTick = input.tick;
    }

    previousTick = input.tick;
  }
}

function validateCompletedTicks(completedTicks: number): void {
  if (
    !Number.isInteger(completedTicks) ||
    completedTicks < 0 ||
    completedTicks > deliveryRushGameRules.totalTicks
  ) {
    throw new RangeError("Completed ticks are outside the game duration.");
  }
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.min(Math.max(value, minimum), maximum);
}

File: backend/beef-and-chicken/beef-and-chicken/Domain/Games/DeliveryRush/DeliveryRushSimulator.cs
namespace beef_and_chicken.Domain.Games.DeliveryRush
{
    public sealed class DeliveryRushSimulator
    {
        public DeliveryRushSimulationResult Simulate(
            int seed,
            IReadOnlyList<DeliveryRushInput> inputs)
        {
            return SimulateUntilTick(
                seed,
                inputs,
                DeliveryRushGameRules.TotalTicks);
        }

        public DeliveryRushSimulationResult SimulateUntilTick(
            int seed,
            IReadOnlyList<DeliveryRushInput> inputs,
            int completedTicks)
        {
            ValidateInputs(inputs);
            ValidateCompletedTicks(completedTicks);

            var random = new DeliveryRushRandom(seed);

            var playerLane =
                DeliveryRushGameRules.StartingLane;

            var inputIndex = 0;

            var nextObstacleTick =
                DeliveryRushGameRules.FirstObstacleTick;

            DeliveryRushObstacle? activeObstacle = null;
            var activeObstacleCollided = false;

            int? jumpStartTick = null;

            var distance = 0;
            var avoidedObstacles = 0;
            var collisionCount = 0;
            var currentCombo = 0;
            var maxCombo = 0;
            var slowdownTicksRemaining = 0;

            var processedTicks = 0;

            for (var tick = 0; tick < completedTicks; tick++)
            {
                processedTicks = tick + 1;

                if (inputIndex < inputs.Count &&
                    inputs[inputIndex].Tick == tick)
                {
                    var input = inputs[inputIndex];

                    switch (input.Action)
                    {
                        case DeliveryRushInput.MoveAction:
                            playerLane = Math.Clamp(
                                playerLane +
                                input.Direction!.Value,
                                0,
                                DeliveryRushGameRules.LaneCount - 1);
                            break;

                        case DeliveryRushInput.JumpAction:
                            jumpStartTick = tick;
                            break;
                    }

                    inputIndex++;
                }

                var distancePerTick =
                    DeliveryRushGameRules
                        .GetDistancePerTick(tick);

                if (slowdownTicksRemaining > 0)
                {
                    distance += Math.Max(
                        1,
                        distancePerTick / 2);

                    slowdownTicksRemaining--;
                }
                else
                {
                    distance += distancePerTick;
                }

                var collisionWindowStartTick =
                    nextObstacleTick -
                    DeliveryRushGameRules.CollisionWindowBeforeTicks;

                if (activeObstacle is null &&
                    tick == collisionWindowStartTick)
                {
                    activeObstacle =
                        DeliveryRushObstacleGenerator.Generate(
                            random,
                            nextObstacleTick);

                    activeObstacleCollided = false;
                }

                var obstacle = activeObstacle;

                if (obstacle is null)
                {
                    continue;
                }

                var collisionWindowEndTick =
                    obstacle.Tick +
                    DeliveryRushGameRules.CollisionWindowAfterTicks;

                var playerLaneMask = 1 << playerLane;

                var isJumping = IsJumpActive(
                    tick,
                    jumpStartTick);

                var isPlayerTouchingObstacle =
                    (obstacle.BlockedLaneMask & playerLaneMask) != 0;

                var jumpAvoidsCollision =
                    isJumping &&
                    DeliveryRushObstacleGenerator.CanJumpOver(
                        obstacle.Type);

                if (!activeObstacleCollided &&
                    isPlayerTouchingObstacle &&
                    !jumpAvoidsCollision)
                {
                    collisionCount++;
                    currentCombo = 0;

                    slowdownTicksRemaining =
                        DeliveryRushGameRules.CollisionSlowdownTicks;

                    activeObstacleCollided = true;

                    if (collisionCount >=
                        DeliveryRushGameRules.MaximumCollisions)
                    {
                        break;
                    }
                }

                if (tick < collisionWindowEndTick)
                {
                    continue;
                }

                if (!activeObstacleCollided)
                {
                    avoidedObstacles++;
                    currentCombo++;

                    maxCombo = Math.Max(
                        maxCombo,
                        currentCombo);
                }

                nextObstacleTick +=
                    DeliveryRushGameRules.GetObstacleInterval(
                        obstacle.Tick);

                activeObstacle = null;
                activeObstacleCollided = false;
            }

            var score =
                distance +
                avoidedObstacles *
                DeliveryRushGameRules.AvoidedObstaclePoints +
                maxCombo *
                DeliveryRushGameRules.MaxComboMultiplier -
                collisionCount *
                DeliveryRushGameRules.CollisionPenalty;

            return new DeliveryRushSimulationResult(
                Score: Math.Max(0, score),
                Distance: distance,
                AvoidedObstacles: avoidedObstacles,
                CollisionCount: collisionCount,
                MaxCombo: maxCombo,
                CompletedTicks: processedTicks);
        }

        public static bool IsJumpActive(
            int currentTick,
            int? jumpStartTick)
        {
            if (!jumpStartTick.HasValue)
            {
                return false;
            }

            return
                currentTick >= jumpStartTick.Value &&
                currentTick <
                jumpStartTick.Value +
                DeliveryRushGameRules.JumpDurationTicks;
        }

        private static void ValidateInputs(
            IReadOnlyList<DeliveryRushInput> inputs)
        {
            ArgumentNullException.ThrowIfNull(inputs);

            if (inputs.Count >
                DeliveryRushGameRules.MaximumInputEvents)
            {
                throw new ArgumentException(
                    "Too many input events.",
                    nameof(inputs));
            }

            var previousTick = -1;
            int? previousJumpTick = null;

            foreach (var input in inputs)
            {
                if (input is null)
                {
                    throw new ArgumentException(
                        "Input cannot be null.",
                        nameof(inputs));
                }

                if (input.Tick < 0 ||
                    input.Tick >=
                    DeliveryRushGameRules.TotalTicks)
                {
                    throw new ArgumentException(
                        "Input tick is outside the game duration.",
                        nameof(inputs));
                }

                if (input.Action is not (
                    DeliveryRushInput.MoveAction or
                    DeliveryRushInput.JumpAction))
                {
                    throw new ArgumentException(
                        "Input action must be move or jump.",
                        nameof(inputs));
                }

                if (input.Action ==
                        DeliveryRushInput.MoveAction &&
                    input.Direction is not (-1 or 1))
                {
                    throw new ArgumentException(
                        "Move direction must be -1 or 1.",
                        nameof(inputs));
                }

                if (input.Action ==
                        DeliveryRushInput.JumpAction &&
                    input.Direction is not null)
                {
                    throw new ArgumentException(
                        "Jump input cannot contain a direction.",
                        nameof(inputs));
                }

                if (input.Tick <= previousTick)
                {
                    throw new ArgumentException(
                        "Input ticks must be strictly increasing.",
                        nameof(inputs));
                }

                if (previousTick >= 0 &&
                    input.Tick - previousTick <
                    DeliveryRushGameRules.MinimumTicksBetweenInputs)
                {
                    throw new ArgumentException(
                        "Input events are too close together.",
                        nameof(inputs));
                }

                if (input.Action ==
                        DeliveryRushInput.JumpAction &&
                    previousJumpTick.HasValue)
                {
                    var nextAllowedJumpTick =
                        previousJumpTick.Value +
                        DeliveryRushGameRules.JumpDurationTicks +
                        DeliveryRushGameRules.JumpCooldownTicks;

                    if (input.Tick < nextAllowedJumpTick)
                    {
                        throw new ArgumentException(
                            "Jump is still on cooldown.",
                            nameof(inputs));
                    }
                }

                if (input.Action ==
                    DeliveryRushInput.JumpAction)
                {
                    previousJumpTick = input.Tick;
                }

                previousTick = input.Tick;
            }
        }

        private static void ValidateCompletedTicks(
            int completedTicks)
        {
            if (completedTicks < 0 ||
                completedTicks >
                DeliveryRushGameRules.TotalTicks)
            {
                throw new ArgumentOutOfRangeException(
                    nameof(completedTicks),
                    "Completed ticks are outside the game duration.");
            }
        }
    }
}